pandas/core/indexes/api.py

```diff
@@ -206,6 +206,17 @@ def conv(i):
 
         return Index(lib.fast_unique_multiple_list([conv(i) for i in inds], sort=sort))
 
+    # GH 35092. Detect if we have an Index type for which the sort
+    # setting doesn't make sense
+    ind_types = list({type(index) for index in indexes})
+    if any(
+        ind_type in [MultiIndex, RangeIndex, DatetimeIndex, CategoricalIndex]
```

Comment:
Please add tests for all of these types that would trigger `ignore_sort=True`

Reply:
@gfyoung 
Re-tested two at a time (old tests were done with a different bugfix implementation, and it confounded the results):

* `MultiIndex` breaks `test_unbalanced` in `ests.reshape.test_melt.TestWideToLong` . From what I can see, we end up not resorting by `id` (first index level) and are instead left with the row order of `year` (the initial `DataFrame`). `MultiIndex` also breaks `test_nonnumeric_suffix` and `test_float_suffix` in `tests.reshape.test_melt.TestWideToLong`. Conceptually, `melt` has the same problem `wide_to_long`.
* `RangeIndex` breaks `test_str_cat_align_mixed_inputs` in `tests.test_strings.TestStringMethods`. There is a indexed/unindexed case and having a default RangeIndex as one of the indices sorts the result. 
* `DatetimeIndex` is safe.
* `CategoricalIndex` breaks `test_construct_with_two_categoricalindex_series` in `tests.frame.test_constructors.TestDataFrameConstructors`. The test assumes that the dataframe resulting from joining two categorical Series will have the categories sorted.

I'm up for changing the tests if you and @jreback think that would be better.